### PR TITLE
Remove implicit interval type coercion from ScalarValue comparison

### PR DESCRIPTION
**Pull request by tustvold**
## Which issue does this PR close?

Part of #7353 

## Rationale for this change

We have explicit type coercion machinery during plan construction that handles this, doing this implicitly within ScalarValue introduces inconsistency and is unnecessary. I am also not sure the type coercion logic is correct, it is certainly not consistent with the arrow-cast kernel.

Potentially more critically this definition of PartialEq is inconsistent with the definition of Hash, which could cause some very strange behaviour

### CODE REVIEW

File: datafusion/common/src/scalar.rs [hunk 1]
```diff
@@ -377,28 +351,10 @@ impl PartialOrd for ScalarValue {
             }
             (TimestampNanosecond(_, _), _) => None,
             (IntervalYearMonth(v1), IntervalYearMonth(v2)) => v1.partial_cmp(v2),
-            (IntervalYearMonth(v1), IntervalDayTime(v2)) => {
-                ym_to_milli(v1).partial_cmp(&dt_to_milli(v2))
```
Comment (tustvold) [inline datafusion/common/src/scalar.rs:381]: This is simply wrong as it makes gross assumptions about the number of days in months, etc...

File: datafusion/common/src/scalar.rs [hunk 2]
```diff
@@ -431,122 +387,6 @@ impl PartialOrd for ScalarValue {
     }
 }
 
-/// This function computes the duration (in milliseconds) of the given
```
Comment (tustvold) [inline datafusion/common/src/scalar.rs:434]: These functions are somewhat misleading as it is not clearly conveyed that they are a gross approximation, assuming certain numbers of days in months, etc... Given these don't appear to be used elsewhere, I think the best option is to remove them

**Review by alamb — APPROVED** [review]
Thank you -- this makes a lot of sense to me. I think some of this was left over from when the arrow cast kernels didn't support all the different types of coercion

cc @berkaysynnada

**Comment by berkaysynnada** [comment]
> Thank you -- this makes a lot of sense to me. I think some of this was left over from when the arrow cast kernels didn't support all the different types of coercion
> 
> cc @berkaysynnada

Yes, you are right. We did this before arrow kernels, when I didn't fully know the purpose of Interval type. I can now say that they were erroneous and they are unnecessary at the moment.

**Pull request merged**